### ISSUE TITLE
Attribute table conditional formatting when using $geometry is NULL works only when reopening the table

### ISSUE THREAD

**Issue by tudorbarascu** [qgis/QGIS#43252]
Attribute table conditional formatting doesn't respect the output of the ``$geometry is NULL`` expression.

**How to Reproduce**
- having a [points.zip](https://github.com/qgis/QGIS/files/6492999/points.zip) layer with two features, one with geometry and one without.
- add a ``$geometry is NULL`` expression as condition (in preview expression it returns 0 for the geometryless feature and 1 for the other as it should)
- both features get highlighted instead of just one which is wrong
![image](https://user-images.githubusercontent.com/3179646/118470772-ac6ab980-b70f-11eb-80af-97dac806769a.png)

Tried also with ``is_empty_or_null($geometry)`` .. same result. Tested on Master (3.19) and on 3.16.7.

**QGIS and OS versions**

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
</head><body>

QGIS version | 3.19.0-Master | QGIS code revision | eaf70803b0
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.2
GDAL/OGR version | 3.2.2
PROJ version | 7.2.1
EPSG Registry database version | v10.008 (2020-12-16)
GEOS version | 3.9.0-CAPI-1.16.2
SQLite version | 3.34.1
PostgreSQL client version | 13.3 (Debian 13.3-1.pgdg110+1)
SpatiaLite version | 5.0.1
QWT version | 6.1.4
QScintilla2 version | 2.11.6
OS version | Debian GNU/Linux 11 (bullseye)
  |   |   |  
Active Python plugins | 

</body></html>
**Additional context**

**Comment by tudorbarascu** [comment]
I don't know why, but for memory layers it works.. very strange. I tried with other formats (shapefile)... doesn't work.

**Comment by elpaso** [comment]
@tudorbarascu works for me on master ....
![immagine](https://user-images.githubusercontent.com/142164/119142272-91b57f00-ba46-11eb-93c4-a1dad622bd2d.png)

**Comment by tudorbarascu** [comment]
@elpaso I can also replicate in:
- another Debian Bullseye VM from installed packages
- under QGIS Master in Ubuntu 20.04 with packages from ``https://qgis.org/ubuntu-nightly focal main``
- under Windows in (3.16.7 and master installed through the osgeo installer).

My normal master build (debian bullseye) is from source and I also tried 3.16.7 from packages.

**Comment by gioman** [comment]
Here (QGIS master from qgis.org repos) on Ubuntu 20.04 I see this:

by default the table only show the record "with_geom", of the formatting rule is set the behavior is ok (the background color is not changed).

the moment I make the table show both records (I have to click "show all features" and "click the refresh button") then the rule is applied to both records.

**Comment by tudorbarascu** [comment]
@gioman Hi Giovanni, should the feedback tag still be available?

**Comment by tudorbarascu** [comment]
Indeed, if the attribute table is opened with the context "show all features", the rule is NOT taken into account but switching forth and back between "Show selected/visible etc" features sometime works.

**Comment by tudorbarascu** [comment]
ping @gioman Hi and sorry for the noise. Just for my information, if we added feedback should the feedback tag still be available?

**Comment by gioman** [comment]
removed, sorry.

**Comment by tudorbarascu** [comment]
Ok, I finally managed to reproduce exactly when it works and when it doesn't work.

It works only after **closing down the attribute table and reopening it**.

Steps to reproduce:

1. Open the Conditional Formating tab and adding the ``$geometry IS NULL`` part and pressing ``Done`` highlights all the features but it supposed to highlight only the **no geometry** feature

2. By closing down the attribute table and reopening it .. the highlight works as it should.

Tested on Master
